lib/ptracer.h

```diff
@@ -0,0 +1,471 @@
+#define PALLENE_TRACER_CONTAINER_ENTRY  "__PALLENE_TRACER_CONTAINER"
+
+/* Finalizer metatable key. */
+/* DO NOT CHANGE EVEN BY MISTAKE. */
+#define PALLENE_TRACER_FINALIZER_ENTRY  "__PALLENE_TRACER_FINALIZER"
+
+/* The size of the Pallene call-stack. */
+/* DO NOT CHANGE EVEN BY MISTAKE. */
+#define PALLENE_TRACER_MAX_CALLSTACK           2000000
+
+/* Traceback ellipsis top threshold. How many frames should we print
+   first to trigger ellipsis? */
+#define PALLENE_TRACEBACK_TOP_THRESHOLD      10
+/* This should always be 2 fewer than top threshold, for symmetry.
+   Becuase we will always have 2 tail frames lingering around at
+   at the end which is not captured by '_countlevels'. Lua also 
+   do it like this. */
+#define PALLENE_TRACEBACK_BOTTOM_THRESHOLD    8
+
+/* ---------------- MACRO DEFINITIONS END ---------------- */
+
+/* ---------------- DATA STRUCTURES ---------------- */
+
+/* What type of frame we are dealing with? Is it just a normal
+   C function or Lua C Function? */
+typedef enum frame_type {
+    PALLENE_TRACER_FRAME_TYPE_C,
+    PALLENE_TRACER_FRAME_TYPE_LUA
+} frame_type_t;
+
+/* Details of the callee function (name, where it is from etc.) */
+/* Optimization Tip: Try making it 'static' for every funciton. */
+typedef struct pt_fn_details {
+    const char *const fn_name;
+    const char *const filename;
+} pt_fn_details_t;
+
+/* A single frame representation. */
+typedef struct pt_frame {
+    frame_type_t type;
+    int line;
+
+    union {
+        pt_fn_details_t *details;
+        lua_CFunction frame_sig;
+    } shared;
+} pt_frame_t;
+
+/* Out stack is fully heap-allocated stack. We need some structure to hold
+   the stack information. This structure will be an Userdatum. */
+/* 'cont' stands for 'container'. */
+typedef struct pt_cont {
+    pt_frame_t *stack;
+    int count;
+} pt_cont_t;
+
+/* ---------------- DATA STRUCTURES END ---------------- */
+
+/* ---------------- DECLARATIONS ---------------- */
+
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+
+/* Pushes a frame to the stack. The frame structure is self-managed for every function. */
+void pallene_tracer_frameenter(lua_State *L, pt_cont_t *cont, pt_frame_t *restrict frame);
+
+/* Set's line number to the topmost frame in the stack. */
+void pallene_tracer_setline(pt_cont_t *cont, int line);
+
+/* Removes the last frame from the stack. */
+void pallene_tracer_frameexit(pt_cont_t *cont);
+
+/* Pallene Tracer explicit traceback function to show Pallene call-stack
+   tracebacks. */
+int  pallene_tracer_debug_traceback(lua_State *L);
+
+/* When we encounter an error, `pallene_tracer_frameexit()` may not
+   get called. Therefore, the stack will get corrupted if the previous
+   call-frames are not removed. The finalizer function makes sure that
+   it does not happen. It's guardian angel. */
+/* The finalizer function will be called from a to-be-closed value (since
+   Lua 5.4). If you are using Lua version prior 5.4, you are outta luck. */
```

Comment:
We should take care not to break completely in older Luas. Can we add some compile-time requirement that Lua version ≥ 5.4 ?

Reply:
Good idea, doing it now.